File: 백준/Gold/9019. DSLR/DSLR.py
import sys
from collections import deque
T = int(input())

for _ in range(T):
    A, B = map(int,sys.stdin.readline().rstrip().split())

    visited = [False for i in range(10001)]
    deq = deque()
    deq.append([A,''])
    visited[A] = True

    while deq:
        num, command = deq.popleft()

        if num == B:
            print(command)
            break

        d = num * 2 % 10000
        if not visited[d]:
            visited[d] = True
            deq.append([d, command + 'D'])

        s = (num - 1) % 10000
        if not visited[s]:
            visited[s] = True
            deq.append([s, command + 'S'])

        l = num // 1000 + (num % 1000)*10
        if not visited[l]:
            visited[l] = True
            deq.append([l, command + 'L'])

        r = num // 10 + (num % 10) * 1000
        if not visited[r]:
            visited[r] = True
            deq.append([r, command + 'R'])

File: 백준/Gold/9019. DSLR/DSLR.js
const fs = require("fs");
const filePath = process.platform === "linux" ? "/dev/stdin" : "input.txt";
const input = fs.readFileSync(filePath).toString().trim().split("\n");

const TC = parseInt(input.shift());
const commands = ["D", "S", "L", "R"];

for (let tc = 0; tc < TC; tc++) {
  const [A, B] = input.shift().split(" ").map(Number);
  console.log(convertAtoB(A, B));
}

function convertAtoB(A, B) {
  let visited = Array.from({ length: 10050 }, () => false);
  let begin = 0;
  let end = 1;
  let queue = [[A, ""]]; //숫자, 계산기록

  while (begin < end) {
    const [num, operation] = queue[begin++];
    if (num == B) return operation;

    for (const c of commands) {
      const newNums = executeCommand(num, c);

      if (!visited[newNums]) {
        visited[newNums] = true;
        queue.push([newNums, operation + c]);
        end++;
      }
    }
  }
}

function executeCommand(number, command) {
  if (command === "D") {
    return (number * 2) % 10000;
  } else if (command === "S") {
    return (number + 9999) % 10000;
  } else if (command === "L") {
    return (number % 1000) * 10 + Math.floor(number / 1000);
  } else if (command === "R") {
    return (number % 10) * 1000 + Math.floor(number / 10);
  }
}
